overlord/ifacestate/helpers.go

```diff
@@ -540,3 +564,127 @@ func resolveSnapIDToName(st *state.State, snapID string) (name string, err error
+	RemapOutgoingSlotRef(slotRef *interfaces.SlotRef) (changed bool)
+}
+
+// NilMapper implements InterfaceMapper and performs no transformations at all.
+type NilMapper struct{}
+
+// RemapIncomingPlugRef doesn't change the plug in any way.
+func (m *NilMapper) RemapIncomingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return false
+}
+
+// RemapOutgoingPlugRef doesn't change the plug in any way.
+func (m *NilMapper) RemapOutgoingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return false
+}
+
+// RemapIncomingSlotRef doesn't change the slot in any way.
+func (m *NilMapper) RemapIncomingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	return false
+}
+
+// RemapOutgoingSlotRef doesn't change the slot in any way.
+func (m *NilMapper) RemapOutgoingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	return false
+}
+
+// mapper contains the currently active interface mapper.
+var mapper InterfaceMapper = &NilMapper{}
+
+// MockInterfaceMapper mocks the currently used interface mapper.
+func MockInterfaceMapper(new InterfaceMapper) (restore func()) {
+	old := mapper
+	mapper = new
+	return func() { mapper = old }
+}
+
+// CoreSnapdMapper implements InterfaceMapper and makes implicit slots
+// appear to be on "core" while they are on "snapd" in reality.
+type CoreSnapdMapper struct{}
+
+// RemapIncomingSlotRef make slots appear on "snapd" rather than "core".
+func (m *CoreSnapdMapper) RemapIncomingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	if slotRef.Snap == "core" {
+		slotRef.Snap = "snapd"
+		return true
+	}
+	return false
+}
+
+// RemapOutgoingSlotRef make slots appear on "core" rather than "snapd".
+func (m *CoreSnapdMapper) RemapOutgoingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	if slotRef.Snap == "snapd" {
+		slotRef.Snap = "core"
+		return true
+	}
+	return false
+}
+
+// RemapIncomingPlugRef doesn't change the plug in any way.
+func (m *CoreSnapdMapper) RemapIncomingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return false
+}
+
+// RemapOutgoingPlugRef doesn't change the plug in any way.
+func (m *CoreSnapdMapper) RemapOutgoingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return false
+}
+
+// Remapping functions for incoming transfers ({state,wire}=>memory)
+
+// RemapIncomingConnRef potentially re-maps connection reference from an API request or being loaded from state.
+//
+// The operation done by remapIncomingConnRef must be symmetric with remapIncomingConnRef.
+// In practice the pair of functions are used to make "snapd" snap the host of implicit
+// interfaces and connections without altering the state in a backwards incompatible way.
+//
+// Data coming from the state and from API requests is changed so that slots on "core"
+// become slots on "snapd" (but only when "snapd" snap itself is being used). When
+// data is about to hit the state again it is re-mapped back.
+func RemapIncomingConnRef(cref *interfaces.ConnRef) (changed bool) {
```

Comment:
I think it would be safer if these Remap* functions didn't do changes in place  but where functional (the internal interface can keep being in place)

Reply:
Done